src-extras/Database/LSMTree/Extras/MergingTreeData.hs

```diff
@@ -0,0 +1,358 @@
+--
+-- Note: 'b ~ Void' should rule out blobs.
+data MergingTreeData k v b =
+    CompletedTreeMergeData (RunData k v b)
+  | OngoingTreeMergeData (MergingRunData MR.TreeMergeType k v b)
+  | PendingLevelMergeData
+      [PreExistingRunData k v b]
+      (Maybe (MergingTreeData k v b))  -- ^ not both empty!
+  | PendingUnionMergeData [MergingTreeData k v b]  -- ^ at least 2 children
+  deriving stock (Show, Eq)
+
+data PreExistingRunData k v b =
+    PreExistingRunData (RunData k v b)
+  | PreExistingMergingRunData (MergingRunData MR.LevelMergeType k v b)
+  deriving stock (Show, Eq)
+
+-- | See @treeInvariant@ in prototype.
+mergingTreeDataInvariant :: MergingTreeData k v b -> Either String ()
+mergingTreeDataInvariant = \case
+    CompletedTreeMergeData _rd ->
+      Right ()
+    OngoingTreeMergeData mr ->
+      mergingRunDataInvariant mr
+    PendingLevelMergeData prs t -> do
+      assertI "pending level merges have at least one input" $
+        length prs + length t > 0
+      for_ prs $ \case
+        PreExistingRunData        _r -> Right ()
+        PreExistingMergingRunData mr -> mergingRunDataInvariant mr
+      for_ t mergingTreeDataInvariant
+    PendingUnionMergeData ts -> do
+      assertI "pending union merges are non-trivial (at least two inputs)" $
+        length ts > 1
+      for_ ts mergingTreeDataInvariant
+  where
+    assertI msg False = Left msg
+    assertI _   True  = Right ()
+
+mapMergingTreeData ::
+     Ord k'
+  => (k -> k') -> (v -> v') -> (b -> b')
+  -> MergingTreeData k v b -> MergingTreeData k' v' b'
+mapMergingTreeData f g h = go
+  where
+    go = \case
+        CompletedTreeMergeData r ->
+          CompletedTreeMergeData $ mapRunData f g h r
+        OngoingTreeMergeData mr ->
+          OngoingTreeMergeData $ mapMergingRunData f g h mr
+        PendingLevelMergeData prs t ->
+          PendingLevelMergeData
+            (map (mapPreExistingRunData f g h) prs)
+            (fmap go t)
+        PendingUnionMergeData ts ->
+          PendingUnionMergeData $ fmap go ts
+
+mapPreExistingRunData ::
+     Ord k'
+  => (k -> k') -> (v -> v') -> (b -> b')
+  -> PreExistingRunData k v b -> PreExistingRunData k' v' b'
+mapPreExistingRunData f g h = \case
+    PreExistingRunData r ->
+      PreExistingRunData (mapRunData f g h r)
+    PreExistingMergingRunData mr ->
+      PreExistingMergingRunData (mapMergingRunData f g h mr)
+
+type SerialisedMergingTreeData =
+    MergingTreeData SerialisedKey SerialisedValue SerialisedBlob
+
+type SerialisedPreExistingRunData =
+    PreExistingRunData SerialisedKey SerialisedValue SerialisedBlob
+
+serialiseMergingTreeData ::
+     (SerialiseKey k, SerialiseValue v, SerialiseValue b)
+  => MergingTreeData k v b -> SerialisedMergingTreeData
+serialiseMergingTreeData =
+    mapMergingTreeData serialiseKey serialiseValue serialiseBlob
+
+{-------------------------------------------------------------------------------
+  QuickCheck
```

Comment:
There is quite a bit of duplication here with the `ScheduledMergesTest` module. Not saying that's a problem, but we could put a TODO somewhere to see if we can reduce duplication, maybe by reusing the `T` type from the prototypes tests?

Reply:
I added a TODO.